pkg/cgroup/cgroup_linux.go

```diff
@@ -0,0 +1,260 @@
+// at /proc/<pid>/cgroup.
+type CGroup struct {
+	ID   int    // Unique cgroup ID (Assigned usually by kernel).
+	Name string // Name of cgroup type.
+	Path string // Path in /sys/fs/cgroup.
+}
+
+// GetCGByName - returns all the cgroups matching the controller name
+// of the cgroup entry.
+func (pcg ProcCGroup) GetCGByName(name string) (res []CGroup) {
+	for _, cg := range pcg {
+		if cg.Name == name {
+			res = append(res, cg)
+		}
+	}
+	return res
+}
+
+// GetProcCGroup - reads and parses all the cgroup entries for a given process.
+func GetProcCGroup(pid int) (ProcCGroup, error) {
+	r, err := os.Open(fmt.Sprintf(procCGroupPath, pid))
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	return parseProcCGroup(r)
+}
+
+// ParseProcCGroup - cgroups are always in the following
+// format once enabled you need to know the pid of the
+// application you are looking for so that the the
+// following parsing logic only parses the file located
+// at /proc/<pid>/cgroup.
+//
+// CGROUP entries id, component and path are always in
+// the following format. ``ID:COMPONENT:PATH``
+//
+// Following code block parses this information and
+// returns a procCGroup which is a parsed list of all
+// the line by line entires from /proc/<pid>/cgroup.
+func parseProcCGroup(r io.Reader) (ProcCGroup, error) {
+	var proc ProcCGroup
+
+	// Start reading cgroup categories line by line
+	// and process them into procCGroup structure.
+	fileScanner := bufio.NewScanner(r)
+	for fileScanner.Scan() {
+		line := fileScanner.Text()
+
+		tokens := strings.SplitN(line, cgroupEntrySep, 3)
+		if len(tokens) < 3 {
+			continue
+		}
+
+		id, name, path := tokens[0], tokens[1], tokens[2]
+		for len(name) != 0 {
+			var next string
+			nameTokens := strings.SplitN(name, cgroupNameSep, 2)
+			if len(nameTokens) == 2 {
+				name = nameTokens[0]
+				next = nameTokens[1]
+			} else {
+				name = nameTokens[0]
+			}
+
+			if strings.HasPrefix(name, cgroupNamePrefix) {
+				name = strings.TrimSpace(name[5:])
+			}
+
+			iid, err := strconv.Atoi(id)
+			if err != nil {
+				return nil, err
+			}
+
+			proc = append(proc, CGroup{
+				ID:   iid,
+				Name: name,
+				Path: path,
+			})
+
+			name = next
+		}
+	}
+
+	// Return upon any error while reading the cgroup categories.
+	if err := fileScanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return proc, nil
+}
+
+// Control group manager CLI defaults.
+const (
+	// Change this if at all in future control group manager
+	// changes their cli name.
+	cgmProgName = "cgm"
+
+	// Action param name used to fetch value for a particular
+	// controller.
+	cgmProgGetValue = "getvalue"
+)
+
+// Fetch value of the cgroup kernel param from the cgroup manager,
+// if cgroup manager is configured we should just rely on `cgm` cli
+// to fetch all the values for us.
+func getCGroupManagerKernValue(cg CGroup, kernParam string) (limit uint64, err error) {
+	limit = unlimitedMemoryLimit
```

Comment:
unlimitedMemoryLimit - is the default value if cgroup is not configured 

```
$ cat /sys/fs/cgroup/memory/memory.limit_in_bytes 
9223372036854771712
```

Just helps in handling the top level code.